includes/class-scheduler.php

```diff
@@ -184,4 +185,96 @@ public static function reprocess_outbox() {
 			self::schedule_outbox_activity_for_federation( $id );
 		}
 	}
+
+	/**
+	 * Asynchronously runs batch processing routines.
+	 *
+	 * The batching part is optional and only comes into play if the callback returns anything.
+	 * Beyond that it's a helper to run a callback asynchronously with locking to prevent simultaneous processing.
+	 *
+	 * @param callable $callback Callable processing routine.
+	 * @params mixed   ...$args  Optional. Parameters that get passed to the callback.
+	 */
+	public static function async_batch( $callback ) {
+		if ( ! \is_callable( $callback ) ) {
+			return;
+		}
+
+		$args = \func_get_args(); // phpcs:ignore PHPCompatibility.FunctionUse.ArgumentFunctionsReportCurrentValue
+		$key  = \md5( \serialize( $callback ) ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+
+		// Bail if the existing lock is still valid.
+		if ( self::is_locked( $key ) ) {
+			\wp_schedule_single_event( time() + MINUTE_IN_SECONDS, 'activitypub_async_batch', $args );
+			return;
+		}
+
+		self::lock( $key );
+
+		$callback = array_shift( $args ); // Remove $callback from arguments.
+		$next     = \call_user_func_array( $callback, $args );
+
+		self::unlock( $key );
+
+		if ( ! empty( $next ) ) {
+			// Schedule the next run, adding the result to the arguments.
+			\wp_schedule_single_event(
```

Comment:
shouldn't we always add the next schedule before we start the federation? otherwise the batch processing might break if one of the batches does not work properly?

Reply:
We could keep track of when the first batch gets started and clear it after the last batch ran, then have `reprocess_outbox` looks for those and pick them back up?